src/main/java/com/amplitude/Retry.java

```diff
@@ -0,0 +1,252 @@
+package com.amplitude;
+
+import org.json.JSONObject;
+
+import java.util.Map;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+class RetryEventsOnceResult {
+  protected boolean shouldRetry;
+  protected boolean shouldReduceEventCount;
+  protected int[] eventIndicesToRemove;
+
+  protected RetryEventsOnceResult(
+      boolean shouldRetry, boolean shouldReduceEventCount, int[] eventIndicesToRemove) {
+    this.shouldRetry = shouldRetry;
+    this.shouldReduceEventCount = shouldReduceEventCount;
+    this.eventIndicesToRemove = eventIndicesToRemove;
+  }
+}
+
+class Retry {
+  // Use map to record the events are currently in retry queue.
+  private static Map<String, Map<String, List<Event>>> idToBuffer =
+      new ConcurrentHashMap<String, Map<String, List<Event>>>();
+  private static int eventsInRetry = 0;
```

Comment:
I guess `eventsInRetry` is not thread safe here, and it can be accessed by multiple threads, should we use `AtomicInteger` instead?